signac/core/file_buffered_synced_collection.py

```diff
@@ -0,0 +1,256 @@
+    def get_buffer_capacity(cls):
+        """Get the current buffer capacity."""
+        return cls._BUFFER_CAPACITY
+
+    @classmethod
+    def set_buffer_capacity(cls, new_capacity):
+        """Update the buffer capacity."""
+        cls._BUFFER_CAPACITY = new_capacity
+
+    @classmethod
+    def get_current_buffer_size(cls):
+        """Get the total amount of data currently stored in the buffer."""
+        return FileBufferedCollection._CURRENT_BUFFER_SIZE
+
+    @staticmethod
+    def _hash(blob):
+        """Calculate and return the md5 hash value for the file data.
+
+        Parameters
+        ----------
+        blob : bytes
+            Byte literal to be hashed.
+
+        Returns
+        -------
+        str
+            The md5 hash of the input bytes.
+        """
+        if blob is not None:
+            m = hashlib.md5()
+            m.update(blob)
+            return m.hexdigest()
+
+    def _get_file_metadata(self):
+        """Return metadata of file.
+
+        Returns
+        -------
+        Tuple[int, float]
+            The size and last modification time of the associated file.
+        """
+        try:
+            metadata = os.stat(self._filename)
+            return metadata.st_size, metadata.st_mtime
+        except OSError as error:
+            if error.errno != errno.ENOENT:
+                raise
+
+    def _flush(self, force=False):
+        """Save buffered changes to the underlying file.
+
+        Parameters
+        ----------
+        force : bool
+            If True, force a flush even in buffered mode (defaults to False).
+        """
+        if not self._is_buffered or force:
+            try:
+                cached_data = self._cache[self._filename]
+            except KeyError:
+                # There are valid reasons for nothing to be in the cache (the
+                # object was never actually accessed during global buffering,
+                # multiple collections pointing to the same file, etc).
+                pass
+            else:
+                blob = json.dumps(self.to_base()).encode()
+
+                # If the contents have not been changed since the initial read,
+                # we don't need to rewrite it.
+                if self._hash(blob) != cached_data["contents"]:
+                    # Validate that the file hasn't been changed by something
+                    # else.
+                    if cached_data["metadata"] != self._get_file_metadata():
+                        raise MetadataError(self._filename)
+                    self._data = json.loads(cached_data["contents"])
+                    self._sync()
+                del self._cache[self._filename]
+                data_size = sys.getsizeof(cached_data)
+                FileBufferedCollection._CURRENT_BUFFER_SIZE -= data_size
+
+    def _encode(self):
+        """Encode the data into a serializable form.
+
+        This method assumes JSON-serializable data, but is exposed to allow
```

Comment:
What do we mean here? This is a private method, it's not really "exposed." If we mean "subclasses can override this method to implement other encoding-decoding schemes" then that can be stated as such.

Reply:
Yes, this is precisely what I am referring to in our conversation on the Redis cache in caching.py. You're right that it's not exposed, I'll reword to indicate that it's a "hook" for subclasses to override serialization behavior. However, there's an open question as to whether or not we like having this restriction here, since it implicitly adds restrictions to any non-JSON SyncedCollection that has a FileBuffered version.

Reply:
That rewording is what I was looking for, sorry if I was unclear. It sounds like we are settling on having that restriction that data must be JSON-serializable here, at least for this iteration of the design.

Reply:
Yes, I think so. Including this hook means that it is possible to use this class as a parent for another object and redefine the data->bytes encoding, so I don't think it's a heavy restriction to default to JSON encoding. @csadorf if you disagree or have a more general solution let us know.